README.md
# RescueOS

RescueOS is a rescue solution for the Nokia N900, distributed as an `initrd`
image.  It has several features, most notably:

- Mounting Maemo `/`
- EMMC access (can mount Maemo home and MyDocs partitions)
- Lock code reset
- WiFi support
- USB mass-storage mode
- USB networking
- Battery charging

Users must be familiar with the Linux console.

This project is *not* the Meego rescue initrd.

# How it works

The flasher utility loads RescueOS.  This makes it unnecessary to modify the
bootloader or Maemo's `/sbin/preinit`, unlike other solutions.

By using the `-l` option, we don't flash the kernel nor the Maemo `initrd`
image.  The kernel is only loaded into RAM.  No modification to the NAND nor
bootloader happens.

# Use cases

Most users don't need it.  That said, this can be used to:

- Fix typos in bootscripts which prevent the OS (Maemo) from booting
- Creating a backup of files before starting a reflash
- Charging the battery (when Maemo is dead)
- Modification of the EMMC partitions and partition table
- Backing up and restoring Maemo rootfs
- `fsck` file system checks
- Resetting the lock code (without reflashing)

# Booting RescueOS

Note: You can't store persistent data in /, since it'ss an initrd.  If you mount
the EMMC or the SD card, they can be written to normally.

**Note**: This comes without any promises and without any warranty.  Therefore,
you're doing everything at your own risk.

To load RescueOS onto an N900, you'll need a USB data cable, a PC, and one of
the following flashers:

- 0xFFFF: the Open Free Fiasco Firmware Flasher
  - Free/open source, but only runs on GNU/Linux
  - Packaged for Ubuntu and Debian as `0xffff`:
    - [Ubuntu](http://packages.ubuntu.com/source/0xffff)
    - [Debian](https://packages.debian.org/sid/0xffff)
  - [Discussion](http://talk.maemo.org/showthread.php?t=87996)
  - [Source](https://github.com/pali/0xFFFF)
- `maemo_flasher` v3.5: the original proprietary flasher from Nokia, for
  Windows, MacOS, and GNU/Linux, but the official distribution site is gone:
  - [Archive](https://web.archive.org/web/20131117084237/http://skeiron.org/tablets-dev/maemo_dev_env_downloads/)
  - [Documentation](http://wiki.maemo.org/Documentation/Maemo_5_Developer_Guide/Development_Environment/Maemo_Flasher-3.5)

Switch the N900 completely off.  If it's plugged into a charger, a PC, or
anything else via the USB port, unplug it and wait about 10 seconds to give the
charger software time to shut down.  Don't connect it to the PC yet.

Start the flasher and get it ready to load RescueOS.  For `0xFFFF`, the command
is:

    0xFFFF -m kernel:rescueOS_n900_kernel_1.3.zImage \
      -m initfs:rescueOS-1.3.img \
      -l -b 'rootdelay root=/dev/ram0'

For `maemo_flasher`, it's:

    flasher-3.5 -k rescueOS_n900_kernel_1.3.zImage \
      -n rescueOS-1.3.img \
      -l -b"rootdelay root=/dev/ram0"

You'll get a message like "suitable device not found..." but the flasher should
block and wait.  Now connect the N900 to the PC with the data cable; the flasher
should detect the device, upload RescueOS, launch it, and then exit.

# How to use it

It boots and you get a bash shell.

The folder `/rescueOS` contains some scripts for various tasks.  Not all of them
are completed yet.  They help you with the usual stuff, e.g. mass-storage
mode, Maemo root mounting, USB networking, WiFi setup, etc.

# Mounting Maemo root

`/rescueOS/mount-maemo-root.sh` mounts the Maemo root to `/mnt/maemo`.
`/rescueOS/umount-maemo-root.sh` unmounts it.

Lock code reset

`code_reset` will set the lock code back to the default (`12345`).

# USB networking

`/rescueOS/usbnetworking-enable.sh` sets up USB networking:

```shell
ifconfig usb0 192.168.2.15 up
ifconfig usb0 netmask 255.255.255.0
route add 192.168.2.14 usb0
```

Disable it with `/rescueOS/usbnetworking-disable.sh`.

# USB mass-storage mode

`/rescueOS/mass-storage-enable.sh` makes the first two EMMC partitions available
for mass-storage mode.  These are respectively `/home` and `/home/user/MyDocs`
inside Maemo.

`/rescueOS/mass-storage-disable.sh` deactivates mass-storage mode.  Remember to
unmount the device on your PC first.

# WiFi support

RescueOS has `wpa_supplicant`, but *without* EAP support.  This should be fine
for most home networks which use a PSK.

If you use a passphrase instead of a passkey, try using:

```shell
wpa_passphrase $ESSID $PASSPHRASE > /run/wlan.conf
sh /rescueOS/setup-wpa-wifi.sh
```

Replace `$ESSID` and `$PASSPHRASE` as appropriate.

For DHCP, use `udhcpc -i wlan0`.

# Battery charging

`/rescueOS/charge21.bash` is based upon ShadowJK's
[charge21](http://enivax.net/jk/n900/) script.  Use it only with a wall charger,
as it pulls 950 mA, and USB isn't designed for that.  If you're sure your USB
port can push an amp and you can live with the consequences of burning out your
USB bus if you're wrong, hey, it's your computer.

# Shutdown

`poweroff`

# Root password

`rootme`

# Shortcuts

The following shortcuts are available to reduce typing on the N900 keyboard.

- `mmr`: Mount Maemo root
- `ummr`: Unmount Maemo root
- `enftp`: Enable FTP (anonymous upload and download everywhere)
- `mse`: Enable mass-storage mode
- `msd`: Disable mass-storage mode
- `usbne`: Enable usb networking
- `usbnd`: Disable usb networking
- `chargebat`: Start battery charging
- `telnetd`: Start `telnetd` (only necessary if you've stopped it explicitly,
  since it starts on boot)

In addition, <kbd>Tab</kbd> can be generated with <kbd>Shift</kbd> + <kbd>Space</kbd> and 
<kbd>Esc</kbd> can be generated with <kbd>Fn</kbd> + <kbd>Ctrl</kbd> (the
<kbd>Fn</kbd> key is marked with <kbd>&#x2197;</kbd>.

# File transfer

Enable the FTP server with `enftp`.  It will have anonymous upload and download
enabled everywhere.  You can connect to it via USB networking or over WiFi,
using any FTP client.

Alternatively, use mass-storage mode.

# Afterword

Contributions and suggestions are welcome.  Write to the original author (NIN101
on freenode or TMO, or `n900-rescueos` at `quitesimple` period `org`).

This fork integrates most of the changes from `gnoutchd` branch of the original,
and includes a change made to the [HACKING](HACKING.md) document made by @gnoutchd
which describes how to verify signatures properly.  Feel free to open an issue
or pull request here if you have any questions.


CHANGELOG
1.3 2016-09-07
=============
FIX	telnetd didn't start automatically

1.2 2015-05-31
==============

NEW	code_reset (can reset the security lock code) (gnoutchd)
NEW	cfdisk, pg, setterm, ul
	(these utilities were left out of previous builds in part because
	 ncurses was disabled at the time that the util-linux package was built) (gnoutchd)
NEW	nano syntax highlighting files updated (gnoutchd)
NEW	kernel rebuilt (gnoutchd)
NEW 	shortcuts so user can type less on the N900 keyboard (NIN101)
NEW 	improved documentation (gnoutchd, NIN101)
REMOVED	depmod, iconv, locale support (gnoutchd)
REMOVED questions during boot (NIN101)
DEVELOP	Kernel and roofs builds now fully automated with buildroot (gnoutchd)

Special thanks to Daniel Gnoutcheff (gnoutchd) for his contributions! 

1.1 2013-06-30
==============
FIX	Correctly detach ubifs (thx, Pali)
FIX	/usr/share/terminfo was missing 
FIX 	Detect device node number of internal emmc on boot, not in mass-storage-enable.sh
NEW	/rescueOS/enabletelnet.sh /rescueOS/enableftp.sh - To enable telnet or ftp manually 

1.0 2013-03-23
==============
NEW	Using buildroot now to create the rootfs.
NEW	updated userspace tools(e. g. busybox).
NEW	glibc replaced with uclibc and everything recompiled against uclibc
NEW	shipping wifi firmware
FIX	corrected dhcp business with /etc/resolv.conf
FIX	/rescueOS/usb-mass-storage-enable.sh: Work if sd card is inserted or not(known mmcblkX number issue on N900)
INFO	decreased size of the initrd
INFO	RescueOS's own configs and scripts are now hosted here:
	https://github.com/NIN101/N900_RescueOS 

0.5.4 2013-01-05 
================
REMOVED midnight commander removed as it was eating too much space and I doubt people were actually using it.
NEW	Now after bootup, the system asks the user whether he wants to enable anonymous ftp up and downloads everywhere.
	Then whether telnetd should be started and finally asks whether usb networking should be enabled.

0.5.3
=====
INFO	it disappeared from the server (some problems with my hoster).
	Development is pretty much stalled. I am focusing on other things.
	Contributions otoh are still welcome.
NEW	Including Pali's script which should fix some messed up maemo /dev/ device nodes (iirc, this can happen with uboot)

0.5.2 2012-03-03
================
Boring and nothing exciting:
NEW	tar is now gnu tar, not busybox's version for convenience
NEW	/tmp/ for some reason never existed, now it does :S
NEW	Mountpoints /mnt/free1, /mnt/free2, /mnt/free3 (how noteworthy!) 
NEW	PS1 was set to show the current directory you are in.

0.5.1 2012-01-22
================
NEW	Battery capacity watchdog tells user at 45 % capacity to charge the 
  	battery. Type "stfu" if the messages annoy you.

0.5 2011-11-02
==============
NEW 	Updated busybox to 1.19.3
NEW 	User can append autostart=/path... option in the kernel 
	command line to run a script right after bootup. 
NEW 	tcpsvd included for ftpd. For example, "tcpsvd -v 0.0.0.0 21 
  	ftpd -w" enables anonymous down- and uploads everywhere. 
NEW 	By Request: Midnight Commander 4.8 included. As functions 
  	keys are not available/mapped, use ESC+number. Still borked. 
NEW cleaner shutdown.

0.4 2011-09-02
==============
NEW	e2fsprogs
NEW	dosfstools
NEW	Now also available as an initrd image.

0.3 2011-08-26
==============
NEW 	Battery charging support. (via ShadowJK's h
  	http://enivax.net/jk/n900/">meegocharge script</a>)

0.2 2011-08-21
==============
INFO 	Decreased rootfs size by a factor of 10
NEW 	First poor-man script to create an archive of the maemo rootfs.¹)
NEW 	Takes the WiFi-Firmwares on first boot from maemo root if wished by 
  	the user.¹
  	fixed. 
NEW 	bash and job control
FIX 	Typo in /rescueOS/umount-maemo-root.sh fixed 
FIX 	Bug which caused the wlan interface number to increase on every reboot is gone.

0.1 2011-07-31
==============
NEW 	First release 
NEW 	WiFi Support
NEW 	ubifs mounting support (maemo root for example)
NEW 	USB networking
NEW 	USB mass-storage mode
NEW 	EMMC mounting



¹ tarball release only, which isn't available anymore. 



rescueOS/fix-maemo-devnodes.sh
#!/bin/sh
#might fix some devices nodes if they have been messed up.
#atm we don't care whether they exist already or not...
#mknod parts by Pali

sh /rescueOS/mount-maemo-root.sh
if [ ! -e /run/maemo-root-mounted ] ; then
echo "mount of maemo root failed, giving up."
exit
fi


mknod /mnt/maemo/dev/console c 5 1
mknod /mnt/maemo/dev/fb0 c 29 0
mknod /mnt/maemo/dev/fb1 c 29 1
mknod /mnt/maemo/dev/fb2 c 29 2
mknod /mnt/maemo/dev/full c 1 7
mknod /mnt/maemo/dev/kmem c 1 2
mknod /mnt/maemo/dev/mem c 1 1
mknod /mnt/maemo/dev/mtd0 c 90 0
mknod /mnt/maemo/dev/mtd0ro c 90 1
mknod /mnt/maemo/dev/mtd1 c 90 2
mknod /mnt/maemo/dev/mtd1ro c 90 3
mknod /mnt/maemo/dev/mtd2 c 90 4
mknod /mnt/maemo/dev/mtd2ro c 90 5
mknod /mnt/maemo/dev/mtd3 c 90 6
mknod /mnt/maemo/dev/mtd3ro c 90 7
mknod /mnt/maemo/dev/mtd4 c 90 8
mknod /mnt/maemo/dev/mtd4ro c 90 9
mknod /mnt/maemo/dev/mtd5 c 90 10
mknod /mnt/maemo/dev/mtd5ro c 90 11
mknod /mnt/maemo/dev/null c 1 3
mknod /mnt/maemo/dev/port c 1 4
mknod /mnt/maemo/dev/random c 1 8
mknod /mnt/maemo/dev/tty c 5 0
mknod /mnt/maemo/dev/ttyS0 c 4 64
mknod /mnt/maemo/dev/ttyS1 c 4 65
mknod /mnt/maemo/dev/ttyS2 c 4 66
mknod /mnt/maemo/dev/tty0 c 4 0
mknod /mnt/maemo/dev/urandom c 1 9
mknod /mnt/maemo/dev/zero c 1 5

mknod /mnt/maemo/dev/loop0 b 7 0
mknod /mnt/maemo/dev/loop1 b 7 1
mknod /mnt/maemo/dev/loop2 b 7 2
mknod /mnt/maemo/dev/loop3 b 7 3
mknod /mnt/maemo/dev/loop4 b 7 4
mknod /mnt/maemo/dev/loop5 b 7 5
mknod /mnt/maemo/dev/loop6 b 7 6
mknod /mnt/maemo/dev/loop7 b 7 7
mknod /mnt/maemo/dev/ram0 b 1 0
mknod /mnt/maemo/dev/ram1 b 1 1
mknod /mnt/maemo/dev/ram10 b 1 10
mknod /mnt/maemo/dev/ram11 b 1 11
mknod /mnt/maemo/dev/ram12 b 1 12
mknod /mnt/maemo/dev/ram13 b 1 13
mknod /mnt/maemo/dev/ram14 b 1 14
mknod /mnt/maemo/dev/ram15 b 1 15
mknod /mnt/maemo/dev/ram16 b 1 16
mknod /mnt/maemo/dev/ram2 b 1 2
mknod /mnt/maemo/dev/ram3 b 1 3
mknod /mnt/maemo/dev/ram4 b 1 4
mknod /mnt/maemo/dev/ram5 b 1 5
mknod /mnt/maemo/dev/ram6 b 1 6
mknod /mnt/maemo/dev/ram7 b 1 7
mknod /mnt/maemo/dev/ram8 b 1 8
mknod /mnt/maemo/dev/ram9 b 1 9

ln -s /proc/kcore /mnt/maemo/dev/core
ln -s ram1 /mnt/maemo/dev/ram

mkdir /mnt/maemo/dev/pts
chmod 755 /mnt/maemo/dev/pts

chmod 600 /mnt/maemo/dev/console
chmod 660 /mnt/maemo/dev/fb0
chmod 660 /mnt/maemo/dev/fb1
chmod 660 /mnt/maemo/dev/fb2
chmod 666 /mnt/maemo/dev/full
chmod 640 /mnt/maemo/dev/kmem
chmod 660 /mnt/maemo/dev/loop0
chmod 660 /mnt/maemo/dev/loop1
chmod 660 /mnt/maemo/dev/loop2
chmod 660 /mnt/maemo/dev/loop3
chmod 660 /mnt/maemo/dev/loop4
chmod 660 /mnt/maemo/dev/loop5
chmod 660 /mnt/maemo/dev/loop6
chmod 660 /mnt/maemo/dev/loop7
chmod 640 /mnt/maemo/dev/mem
chmod 600 /mnt/maemo/dev/mtd0
chmod 600 /mnt/maemo/dev/mtd0ro
chmod 600 /mnt/maemo/dev/mtd1
chmod 600 /mnt/maemo/dev/mtd1ro
chmod 600 /mnt/maemo/dev/mtd2
chmod 600 /mnt/maemo/dev/mtd2ro
chmod 600 /mnt/maemo/dev/mtd3
chmod 600 /mnt/maemo/dev/mtd3ro
chmod 600 /mnt/maemo/dev/mtd4
chmod 600 /mnt/maemo/dev/mtd4ro
chmod 600 /mnt/maemo/dev/mtd5
chmod 600 /mnt/maemo/dev/mtd5ro
chmod 666 /mnt/maemo/dev/null
chmod 640 /mnt/maemo/dev/port
chmod 660 /mnt/maemo/dev/ram0
chmod 660 /mnt/maemo/dev/ram1
chmod 660 /mnt/maemo/dev/ram10
chmod 660 /mnt/maemo/dev/ram11
chmod 660 /mnt/maemo/dev/ram12
chmod 660 /mnt/maemo/dev/ram13
chmod 660 /mnt/maemo/dev/ram14
chmod 660 /mnt/maemo/dev/ram15
chmod 660 /mnt/maemo/dev/ram16
chmod 660 /mnt/maemo/dev/ram2
chmod 660 /mnt/maemo/dev/ram3
chmod 660 /mnt/maemo/dev/ram4
chmod 660 /mnt/maemo/dev/ram5
chmod 660 /mnt/maemo/dev/ram6
chmod 660 /mnt/maemo/dev/ram7
chmod 660 /mnt/maemo/dev/ram8
chmod 660 /mnt/maemo/dev/ram9
chmod 666 /mnt/maemo/dev/random
chmod 666 /mnt/maemo/dev/tty
chmod 660 /mnt/maemo/dev/ttyS0
chmod 660 /mnt/maemo/dev/ttyS1
chmod 660 /mnt/maemo/dev/ttyS2
chmod 600 /mnt/maemo/dev/tty0
chmod 444 /mnt/maemo/dev/urandom
chmod 666 /mnt/maemo/dev/zero

chown root:tty /mnt/maemo/dev/console
chown root:video /mnt/maemo/dev/fb0
chown root:video /mnt/maemo/dev/fb1
chown root:video /mnt/maemo/dev/fb2
chown root:root /mnt/maemo/dev/full
chown root:kmem /mnt/maemo/dev/kmem
chown root:disk /mnt/maemo/dev/loop0
chown root:disk /mnt/maemo/dev/loop1
chown root:disk /mnt/maemo/dev/loop2
chown root:disk /mnt/maemo/dev/loop3
chown root:disk /mnt/maemo/dev/loop4
chown root:disk /mnt/maemo/dev/loop5
chown root:disk /mnt/maemo/dev/loop6
chown root:disk /mnt/maemo/dev/loop7
chown root:kmem /mnt/maemo/dev/mem
chown root:root /mnt/maemo/dev/mtd0
chown root:root /mnt/maemo/dev/mtd0ro
chown root:root /mnt/maemo/dev/mtd1
chown root:root /mnt/maemo/dev/mtd1ro
chown root:root /mnt/maemo/dev/mtd2
chown root:root /mnt/maemo/dev/mtd2ro
chown root:root /mnt/maemo/dev/mtd3
chown root:root /mnt/maemo/dev/mtd3ro
chown root:root /mnt/maemo/dev/mtd4
chown root:root /mnt/maemo/dev/mtd4ro
chown root:root /mnt/maemo/dev/mtd5
chown root:root /mnt/maemo/dev/mtd5ro
chown root:root /mnt/maemo/dev/null
chown root:kmem /mnt/maemo/dev/port
chown root:disk /mnt/maemo/dev/ram0
chown root:disk /mnt/maemo/dev/ram1
chown root:disk /mnt/maemo/dev/ram10
chown root:disk /mnt/maemo/dev/ram11
chown root:disk /mnt/maemo/dev/ram12
chown root:disk /mnt/maemo/dev/ram13
chown root:disk /mnt/maemo/dev/ram14
chown root:disk /mnt/maemo/dev/ram15
chown root:disk /mnt/maemo/dev/ram16
chown root:disk /mnt/maemo/dev/ram2
chown root:disk /mnt/maemo/dev/ram3
chown root:disk /mnt/maemo/dev/ram4
chown root:disk /mnt/maemo/dev/ram5
chown root:disk /mnt/maemo/dev/ram6
chown root:disk /mnt/maemo/dev/ram7
chown root:disk /mnt/maemo/dev/ram8
chown root:disk /mnt/maemo/dev/ram9
chown root:root /mnt/maemo/dev/random
chown root:tty /mnt/maemo/dev/tty
chown root:dialout /mnt/maemo/dev/ttyS0
chown root:dialout /mnt/maemo/dev/ttyS1
chown root:dialout /mnt/maemo/dev/ttyS2
chown root:tty /mnt/maemo/dev/tty0
chown root:root /mnt/maemo/dev/urandom
chown root:root /mnt/maemo/dev/zero


rescueOS/charge21.bash
#/bin/bash
#Original by ShadowJK. http://enivax.net/jk/n900/
if [ -e /run/charger-runned ] ; then
echo "Script already started. Delete run/charger-runned if you are sure no other instance is running"
exit
fi
touch /run/charger-runned
echo "Charger: " $(cat /sys/devices/platform/musb_hdrc/charger)

I2CGET=/usr/bin/i2cget
I2CSET=/usr/bin/i2cset

if pgrep bme_RX-51; then
	echo "BME is running. Exiting."
	exit
fi

configure()
{
  echo -n "Pre-Config Status: " $($I2CGET -y 2 0x6b 0x00)
  echo -n ", " $($I2CSET -y 2 0x6b 0x00) "."

  # Disable charger for configuration:
  # $I2CSET -y 2 0x6b 0x01 0xcc # No limit, 3.4V weak threshold, enable term, charger disable

  # Register 0x04
  # 8: reset
  # 4: 27.2mV  # charge current
  # 2: 13.6mV
  # 1: 6.8mV
  # 8: N/A
  # 4: 13.6mV # termination current
  # 2: 6.8mV
  # 1: 3.4mV
  # 7-1250 6-1150 5-1050 4-950 3-850 2-750 1-650 0-550
  # 7-400 6-350 5-300 4-250 3-200 2-150 1-100 0-50
  $I2CSET -y -m 0xFF 2 0x6b 0x04 0x42;

  # Register 0x02
  # 8: .640 V
  # 4: .320 V
  # 2: .160 V
  # 1: .080
  # 8: .040
  # 4: .020 (+ 3.5)
  # 2: otg pin active at high (default 1)
  # 1: enable otg pin
  $I2CSET -y -m 0xfc 2 0x6b 0x02 0x8c;
  REGV=4200
  # 4.24 = 3.5 + .640 + .08 + .02 = 94
  # 4.22 = 3.5 + ,640 + .08 = 90
  # 4.2 = 3.5 + .640 + .040 + .02 = 8c
  # 4.16 = 3.5 + .640V + .020 = 84
  # 4.1 = 3.5 + .320 + .160 + .08 + .04 = 78
  # 4.0 = 3.5 + .320 + .160 + .02 = 64
  # 3.9 = 3.5 + .320 + .080 = 50

  # Register 0x1
  # 8: 00 = 100, 01 = 500, 10 = 800mA
  # 4: 11 = no limit
  # 2: 200mV weak threshold default 1
  # 1: 100mV weak treshold defsult 1 (3.4 - 3.7)
  # 8: enable termination
  # 4: charger disable
  # 2: high imp mode
  # 1: boost
  $I2CSET -y 2 0x6b 0x01 0x08;
  #hotswap: c0
  #normal: c8


  # Register 0x00
  # 8: Read:  OTG Pin Status
  #    Write: Timer Reset
  # 4: Enable Stat Pin
  # 2: Stat : 00 Ready 01 In Progress
  # 1:      : 10 Done  11 Fault
  # 8: Boost Mode
  # 4: Fault: 000 Normal 001 VBUS OVP 010 Sleep Mode
  # 2:        011 Poor input or Vbus < UVLO
  # 1:        100 Battery OVP 101 Thermal Shutdown
  #           110 Timer Fault 111 NA
  $I2CSET -y 2 0x6b 0x00 0x00;

  # Softstart

  $I2CSET -y 2 0x6b 0x01 0xc8;
  sleep 1
  #echo " Unlimited: " $($I2CGET -y 2 0x6b 0x00)
  #$I2CSET -y 2 0x6b 0x00 0x80 # timer reset
  cat /sys/devices/platform/musb_hdrc/charger >/dev/null
}

configure

# Initialize variables
FULL=0
MODE="STANDBY"
WALLCHARGER=0

# Assuming a nice round number 20mOhm for bq27200 sense resistor
RS=21

# Assuming this much resistance between charger and battery
CR=131

BQPATH="/sys/class/power_supply/bq27200-0"

get_nac ()
{
    NAC=$(cat $BQPATH/charge_now)
    NAC=$((NAC/1000))
}
get_rsoc ()
{
    RSOC=$(cat $BQPATH/capacity)
}
get_volt ()
{
   VOLT=$(cat $BQPATH/voltage_now)
   VOLT=$((VOLT/1000))
}
get_rate ()
{
    RATE=$(cat $BQPATH/current_now)
    RATE=$((RATE/1000))
    RATE=$((-RATE))
}
calculate_system_use ()
{
    SYSTEM_USE=$(( ($REGV *1000 - $VOLT*1000) / $CR - $RATE ))
    if [ $(($SYSTEM_USE+$RATE)) -gt 950 ] ; then
        SYSTEM_USE=$((950-RATE))
    fi
}
decode_status()
{
  S=$1
  SOTG=$((S/128))
  S=$((S-SOTG*128))

  SSPIN=$((S/64))
  S=$((S-SSPIN*64))

  SSTAT=$((S/16))
  S=$((S-SSTAT*16))

  SBOOST=$((S/8))
  S=$((S-SBOOST*8))

  SFAULT=$((S))

#  if [ $SSTAT -eq 0 ] ; then
#    echo charger status: Ready
#  fi
#  if [ $SSTAT -eq 1 ] ; then
#    echo charger status: InProgress
#  fi
#  if [ $SSTAT -eq 2 ] ; then
#    echo charger status: Done
#  fi
  if [ $SSTAT -eq 3 ] || [ $SFAULT -ne 0 ]; then
    echo -n $(date) charger status: Fault:
    case $SFAULT in
     0 )
       echo NoFault  ;;
     1 )
       echo VbusOVP ;;
     2 )
       echo SleepMode ;;
     3 )
       echo Poor Input or VBus UnderVoltage ;;
     4 )
       echo Battery Overvoltage Protection ;;
     5 )
       echo Thermal Shutdown ;;
     6 )
       echo TimerFault ;;
     7 )
       echo Unknown ;;
    esac
  fi
}


SLEEPTIME=15

while true ; do
   STATUS=$($I2CGET -y 2 0x6b 0x00)
   #echo $STATUS

   $I2CSET -y -m 0x80 2 0x6b 0x00 0x80; # timer reset
   get_nac
   get_rsoc
   get_volt
   get_rate

   # Sanity
   if [ $VOLT -gt 4200 ] ; then
      echo "***CRITICAL*** Battery voltage $VOLT exceeds 4.2V!!!"
      exit 2
   fi

   if [ $MODE == "STANDBY" ] ; then
      if [ $STATUS == 0x10 ] || [ $STATUS == 0x90 ] ; then
         MODE="CHARGING" ; SLEEPTIME=5
         echo $(date) "standby -> CHARGING. Current available capacity: " $NAC "mAh, " $RSOC " percent."
         WALLCHARGER=$(cat /sys/devices/platform/musb_hdrc/charger)
      fi
   fi

   if [ $MODE == "CHARGING" ] ; then
      if [ $STATUS == 0x00 ] ; then
         MODE="STANDBY" ; SLEEPTIME=15
         echo $(date) "charging -> STANDBY. Current available capacity: " $NAC "mAh, " $RSOC " percent."
         WALLCHARGER=0
         # This will stop USB from eating power as long as you haven't plugged it into a PC
#       ####  echo 0 > /sys/devices/platform/musb_hdrc/connect
      fi
   fi

   if [ $STATUS == 0x20 ] && [ $FULL == 0 ] ; then
      echo "Charge done"
      #echo $(date) "FULL: " $NAC "mAh" >> /home/user/MyDocs/charger.log
      FULL=1
   fi

   if [ $STATUS == 0x00 ] && [ $FULL == 1 ] ; then
      FULL=0
   fi

   calculate_system_use
   decode_status $STATUS
   echo Status: $STATUS Mode: $MODE Full: $FULL Wall: $WALLCHARGER Voltage: $VOLT NAC: $NAC level: $RSOC % Rate: $RATE System: "-"$SYSTEM_USE Ch: $(($SYSTEM_USE+$RATE))
   sleep $SLEEPTIME
done
